lib/tts_isolate.dart

```diff
@@ -52,44 +56,29 @@ void onStart(ServiceInstance service) async {
   DartPluginRegistrant.ensureInitialized();
   WidgetsFlutterBinding.ensureInitialized();
   await Firebase.initializeApp();
+  // final prefs = await StreamingSharedPreferences.instance;
```

Comment:
We should listen on the prefs instead of the tts channel passed through the argument. So in order to start tts, we should do two things:

1. Call `FlutterBackgroundService().invoke('startTts');` like you're doing now, but don't pass the channel.
2. Update shared prefs `tts_channel` to the tts channel json.

This is so we can update the channel regardless of if the service is started or not. Otherwise, right now you can't switch channels with tts on (well, you can but tts won't carry over).

Reply:
I had one issue attempting to persist data from `prefs` from main isolate to tts isolate. Initializing the TTS would actually update the value I had tested with from false to true, but disabling it, no matter how many times I tried, the data would stay `true` and never kill the listen service. 

I have a lot less knowledge on this than you, but to my understanding each isolate has it's own cache. I am assuming there is a way to work around this, maybe with a singleton class or something? What do you suggest? 

Also would it be better to stop using `streaming_shared_preferences` and use `shared_preferences` instead as it looks like the former was abandoned some years ago. 

Here is what I've [found](https://stackoverflow.com/a/75258254/13284500) on the subject, let me know your thoughts before I continue here.

Reply:
Interesting, well that fuckin' sucks. Ok, so then let's try to use the `invoke` in a slightly different way: instead of `invoke("startTts", ...)` let's make it just `invoke("setTtsChannel", ...);` and if you want to stop TTS, do `invoke("setTtsChannel", { channel: null });`.

In the isolate, we should listen on `setTtsChannel` and if we receive a null, stop listening and exit the thread. This would look something like:

```
Subscription? messagesSubscription = null;
Subscription? channelSubscription = bgchannel.on("setTtsChannel").listen((event) => {
  if (event.channel == null) {
    channelSubscription.unsubscribe();
  }.else {
    messagesSubscription?.unsubscribe();
    messagesSubscription = Firestore.collection(...).snapshots().listen((data) => {
      // read message
    });
  }
});
```

This should exit the isolate when `channel: null` is sent.

Reply:
Re: removing `streaming_shared_preferences` yeah that sounds reasonable if we do the above approach. I had hoped `streaming_shared_preferences` would work as a way to communicate between the isolates but I guess not. :(

Reply:
If you can take a look at the latest commit I did based on your response to see if it's what you'd like. Let me know if there needs to be any refactoring. I tested it, it's working as well as the other way I had it. 

A couple things I was unsure of, if it was safe to put 
```
StreamSubscription? messagesSubscription;
StreamSubscription? channelSubscription;
```
at the top of the file like I did, or if it was okay to be be in the function. 

Also you had it this way, 
```dart
  if (event.channel == null) {
    channelSubscription.unsubscribe();
  }.else {
  ```
  
  To make it work it had to be like below. If I canceled the `channelSubscription` it would not allow it to work again when re enabling the TTS.
  ```dart
  if (args['channel'] == null) {
        await ttsQueue.clear();
        messagesSubscription?.cancel();
      } else {
      ```
Just wanted to make sure that was alright?

Reply:
Also completely unrelated, this doesn't happen outside of the TTS, but might be worth looking into or it might be safe to ignore. But when you disable TTS (regardless of the isolate code) you get this warning. 

![image](https://github.com/muxable/rtchat/assets/48299282/3eeeb377-dbca-4ced-a9cd-fdc92b60fa85)